### ISSUE TITLE
How to generate the tfrecord files for other datasets?

### ISSUE THREAD

**Issue by lzh19961031** [catarina-barata/skin-hierarchy#2]
Could the author provide the code to generate tfrecored file  for other datasets, such as ISIC2016, PH2? Since the structure of the tfrecord is unknown.

**Comment by catarina-barata** [comment]
Sorry for the delay - out of the office.

Please adapt the following code to adjust to your data taxonomy and image names. 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%5
# -*- coding: utf-8 -*-
"""
Spyder Editor

This is a temporary script file.
"""

# -*- coding: utf-8 -*-
"""
Created on Fri Feb 23 12:39:00 2018

@author: ist
"""
from PIL import Image
import numpy as np

import tensorflow.compat.v1 as tf
tf.disable_v2_behavior()

import os.path
import skimage.transform as trans


image_size = 401


def _bytes_feature(value):
    return tf.train.Feature(bytes_list=tf.train.BytesList(value=[value]))

def _int64_feature(value):
    return tf.train.Feature(int64_list=tf.train.Int64List(value=[value]))

tfrecords_filename = 'something.tfrecords' #Your directory/tfrecord_name.tfrecords

Token = 'START'

# Let's collect the real images to later on compare
# to the reconstructed ones
original_images = []

img_dir = '' #Your image directory - should be organized in folders and subfolders that convey hierarchy

directories = [d for d in os.listdir(img_dir) if os.path.isdir(os.path.join(img_dir, d))]
labels = []
image_file = []

images = []
g = 0
writer = tf.python_io.TFRecordWriter(tfrecords_filename)

for d in directories:
    
    with tf.device('/cpu:0'):
        label_directory = os.path.join(img_dir,d)
        
        sub_dir = [d for d in os.listdir(label_directory) if os.path.isdir(os.path.join(label_directory, d))]
        
       
        for s in sub_dir:
            
            if len( [f for f in os.listdir(os.path.join(label_directory, s))]) < 10:
                
                sub_sub_dir = [f for f in os.listdir(os.path.join(label_directory, s))]
                
                for sub in sub_sub_dir:
                    file_names = [os.path.join(os.path.join(label_directory, s),f) for f in os.listdir(os.path.join(label_directory, s)) if not f.endswith("_norm.jpg")]
                    file_names = [os.path.join(os.path.join(os.path.join(label_directory, s),sub),f) for f in os.listdir(os.path.join(os.path.join(label_directory, s),sub)) if not f.endswith("_crop_norm.jpg")]

                    
                    caption = Token+ ' '+d+' '+s+' '+sub+' END'
                    it = 0
                    #caption = d+' '+s+' .'
                    for f in file_names:
                            
                        img = np.array(Image.open(f))
                        
                        img = np.asarray(trans.resize(img,[image_size,image_size,3],preserve_range=True),dtype=np.uint8)
                        
                        height = img.shape[0]
                        width = img.shape[1]
                            
                        img_raw = img.tostring()
                            
                        example = tf.train.Example(features=tf.train.Features(feature={
                            'height': _int64_feature(height),
                            'width': _int64_feature(width),
                            'image_raw': _bytes_feature(tf.compat.as_bytes(img_raw)),
                            'file': _bytes_feature(tf.compat.as_bytes(f)),
                            'labels': _bytes_feature(tf.compat.as_bytes(caption))}))
                            
                        writer.write(example.SerializeToString())
                        
                        it+=1
            else:
                
                file_names = [os.path.join(os.path.join(label_directory, s),f) for f in os.listdir(os.path.join(label_directory, s))]
      
                files = [f for f in os.listdir(os.path.join(label_directory, s))]
                
                caption = Token+ ' '+d+' '+s+' END'
                it = 0
                caption = d+' '+s+' .'
                for f in file_names:
                        
                    img = np.array(Image.open(f))
                    
                    img = np.asarray(trans.resize(img,[image_size,image_size,3],preserve_range=True),dtype=np.uint8)
                    
                    height = img.shape[0]
                    width = img.shape[1]
                        
                    img_raw = img.tostring()
                        
                    example = tf.train.Example(features=tf.train.Features(feature={
                        'height': _int64_feature(height),
                        'width': _int64_feature(width),
                        'image_raw': _bytes_feature(tf.compat.as_bytes(img_raw)),
                        'file': _bytes_feature(tf.compat.as_bytes(files[it])),
                        'labels': _bytes_feature(tf.compat.as_bytes(caption))}))
                        
                    writer.write(example.SerializeToString())
                    
                    it+= 1

writer.close()

with tf.device('/cpu:0'):
    record_iterator = tf.python_io.tf_record_iterator(path=tfrecords_filename)
    
    x=0
    for string_record in record_iterator:
        # Parse the next example
        example = tf.train.Example()
        example.ParseFromString(string_record)
    
        # Get the features you stored (change to match your tfrecord writing code)
        height = int(example.features.feature['height']
                                     .int64_list
                                     .value[0])
    
        width = int(example.features.feature['width']
                                    .int64_list
                                    .value[0])
    
        img_string = (example.features.feature['image_raw']
                                      .bytes_list
                                      .value[0])
        
        l = (example.features.feature['labels'].bytes_list.value[0])
        
        file = (example.features.feature['file'].bytes_list.value[0])
        
        # Convert to a numpy array (change dtype to the datatype you stored)
        img_1d = np.fromstring(img_string, dtype=np.uint8)
        # Print the image shape; does it match your expectations?
        print(img_1d.shape)
        print(l.decode())
        print(file.decode())
        x+=1
    print(x)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
